state/txindex/kv/kv.go

```diff
@@ -75,18 +75,30 @@ func (txi *TxIndex) Get(hash []byte) (*types.TxResult, error) {
 	return txResult, nil
 }
 
-// AddBatch indexes a batch of transactions using the given list of tags.
+// AddBatch indexes a batch of transactions using the given list of events. Each
+// key that indexed from the tx's events is a composite of the event type and
+// the respective attribute's key delimited by a "." (eg. "account.number").
+// Any event with an empty type is not indexed.
 func (txi *TxIndex) AddBatch(b *txindex.Batch) error {
 	storeBatch := txi.store.NewBatch()
 	defer storeBatch.Close()
 
 	for _, result := range b.Ops {
 		hash := result.Tx.Hash()
 
-		// index tx by tags
-		for _, tag := range result.Result.Tags {
-			if txi.indexAllTags || cmn.StringInSlice(string(tag.Key), txi.tagsToIndex) {
-				storeBatch.Set(keyForTag(tag, result), hash)
+		// index tx by events
+		for _, event := range result.Result.Events {
+			// only index events with a non-empty type
+			if len(event.Type) > 0 {
+				for _, attr := range event.Attributes {
+					if len(attr.Key) > 0 {
+						compositeTag := fmt.Sprintf("%s.%s", event.Type, string(attr.Key))
+						if txi.indexAllTags || cmn.StringInSlice(compositeTag, txi.tagsToIndex) {
+							storeBatch.Set(keyForEvent(compositeTag, attr.Value, result), hash)
```

Comment:
Ah. This means the change is breaking for the indexer. Not sure if there's a way to avoid that (prob not), which means we may need to re-index the tx db on startup or else somehow prevent queries against the older data ... (probably want the former)

Reply:
I'm not necessarily sure that old queries will break or no longer work -- I think they should, but I have yet to test this. The key format is still `key/value/height/index`.

Reply:
But if txs have already been indexed under the old format, then they won't be accessible under the new format. Ie if an existing node upgrades, won't queries for older heights fail since the key is now a composite?

Reply:
Hmmm, so just spitballing here, but I don't immediately see why queries wont work for older heights (pre-upgrade) as the format _structure_ remains unchanged -- on the key is changed.

Say for example in the context of the SDK/Gaia, we emit a tag `"sender": "cosmos0xxx"` and now is updated  to emit an equivalent event `"transfer": [{"sender": "cosmos0xxx"}]`, the old tag would be indexed under `sender/cosmos0xxx/H/I` and the new event would be indexed under `transfer.sender/cosmos0xxx/H/I`.

Now when we search via `*TxIndex#Search`, both queries `sender = "cosmos0xxx"` and `transfer.sender = "cosmos0xxx"` should work as we just call `startKeyForCondition` for the given `Condition` which in turn does a prefix lookup with the give bytes from `startKeyForCondition`.

Maybe I'm completely wrong here and nodes will have to resync...

Reply:
> Hmmm, so just spitballing here, but I don't immediately see why queries wont work

Maybe we should add a small test to ensure?

Reply:
I've implemented a test `TestTxSearchDeprecatedIndexing` that indexes in the deprecated fashion which shows querying still works on old txs/keys.